### DISCUSSION TITLE
nextjs withRouter query object issue

### DISCUSSION THREAD

**Discussion by EhsanKhalili** [vercel/next.js#17276] [Help] [answered]
# Bug report

withRouter query object return null

## Describe the bug

in my component I used WithRouter and I want to pull query string object from it, but at the first render(I mean for example refresh the page), router.query return an empty object, but my url is something like : .../question?id=sdsdsdsadda

**Comment by EhsanKhalili** [comment]
it's a simple project generated with create-next-app,

import React from 'react';
import { withRouter } from 'next/router';

class Test extends React.Component {
   componentDidMount() {
      const { router } = this.props;
      console.log(router.query);
   }

   render() {
      return <div>TEST</div>;
   }
}

export default withRouter(Test);

when I open http://localhost:3000/test?ehsan=2, console will log an empty objecy: {}

**Comment by jamesmosier** [comment, marked as answer]
That is to be expected. On the first render the router is not hydrated yet. Therefore you must account for this in your own code and wait for `router` to have a value.

See [this discussion answer](https://github.com/vercel/next.js/discussions/14657#discussioncomment-31264) for more info or the [docs](https://nextjs.org/docs/routing/dynamic-routes#caveats) mention this too.
